### PR TITLE
Feature tables3 testing

### PR DESCRIPTION
**Pull request by ian-r-rose**
This does two major things:
1. Implements the cemetery for listfields and textfields. There are some reasonably significant changes to the implementation of patching for those fields, so some scrutiny there would be good.
2. Adds a test suite. There are always more to add, but I think the current set is reasonably good, and could use some review.

### CODE REVIEW

**Comment by saulshanabrook** [comment]
I reverted the coverage support. It might have to do with a differering node version on Travis and I didn't want to update it in this PR in case it broke other things. From the logs:

> 18 07 2019 23:38:10.785:ERROR [plugin]: Error during loading "/home/travis/build/phosphorjs/phosphor/node_modules/karma-coverage-istanbul-reporter" plugin:
  Block-scoped declarations (let, const, function, class) not yet supported outside strict mode

**Comment by ian-r-rose** [comment]
Sounds good, thanks @saulshanabrook. All green now!

File: packages/datastore/src/datastore.ts [hunk 1]
```diff
@@ -596,6 +637,24 @@ namespace Datastore {
...
   }
 
+  /**
+   * A message of a datastore GC opportunity.
+   *
+   * You can send such a message to the datastore when you
+   * are sure that all concurrent transactions have been applied,
+   * i.e. any transactions that arrive later will have the current
+   * state as a base. This will provide  a hint to the datastore
+   * that it is safe to clean up some internal state.
```
Comment (vidartf) [inline packages/datastore/src/datastore.ts]: This note might not be correct given my misunderstanding which models we were using (causal delivery not being a requirement). Somebody will have to check:
- Is this correct? If not:
- Are there ever any time that is better for GC than others?

File: packages/datastore/src/listfield.ts [hunk 2]
```diff
@@ -525,23 +539,50 @@ namespace Private {
...
+
       // Set up the chunk ids.
       let chunkIds: string[] = [];
 
       // Set up the chunk values.
       let chunkValues: T[] = [];
 
-      // Find the extent of the chunk.
-      while (i < n && StringExt.cmp(ids[i], metadataIds[j]) < 0) {
+      // Bail early if the boundary is the end of the array.
```
Comment (sccolbert) [inline packages/datastore/src/listfield.ts]: Can you explain a bit what is going on here? I'm wondering if we can simplify this in any way.

File: packages/datastore/src/schema.ts [hunk 3]
```diff
@@ -46,7 +46,7 @@ function validateSchema(schema: Schema): string[] {
   const errors = [];
   // Ensure that field names do not begin with `$` or `@`.
   for (let name in schema.fields) {
-    if (invalidFielnameLeads.indexOf(name[0]) !== -1) {
+    if (invalidFieldnameLeads.indexOf(name[0]) !== -1) {
```
Comment (sccolbert) [inline packages/datastore/src/schema.ts]: why `indexOf` and not simply `===`?

Comment (ian-r-rose) [reply packages/datastore/src/schema.ts]: Not sure, I was just correcting a typo here.

File: packages/datastore/src/listfield.ts (hunk 2 above)
Comment (ian-r-rose) [reply packages/datastore/src/listfield.ts]: Essentially, we have to check the cemetery for *every* inserted metadata ID to see if it should be added or not. So if we are inserting a chunk, it's not enough to just check the leading ID.

So the changed logic here is to check all every the cemetery for every ID. The actual inserted chunks that result from this function should not be different, except that they will be filtered of IDs that are in the cemetery.

**Pull request merged**